File: frontend/src/api/recordsApi.js
// Record API functions bound to the HEIRS Spring Boot REST contract.
// All endpoint shapes follow backend/API.md exactly. Records are normalized
// here so components receive the frontend data shape (e.g. keyword arrays).
import { apiFetch } from './apiClient';

export const MAX_PAGE_SIZE = 100;

// Frontend category aliases -> backend enum label. The approved filter UI
// uses "Rule" while the backend enum is "Rules", so this mapping is required.
const CATEGORY_TO_API = {
  Regulation: 'Regulation',
  Policy: 'Policy',
  Project: 'Project',
  Rule: 'Rules',
  Rules: 'Rules',
  Scheme: 'Scheme',
};

export function categoryToApi(value) {
  if (!value) return '';
  const key = String(value).trim();
  return CATEGORY_TO_API[key] || key;
}

function keywordsToArray(keywords) {
  if (!keywords) return [];
  if (Array.isArray(keywords)) return keywords.map((k) => String(k).trim()).filter(Boolean);
  return String(keywords)
    .split(',')
    .map((keyword) => keyword.trim())
    .filter(Boolean);
}

// Convert a backend record JSON object into the frontend record shape.
export function normalizeRecord(record) {
  if (!record) return record;
  return {
    ...record,
    keywords: keywordsToArray(record.keywords),
  };
}

function normalizeList(payload) {
  if (!payload || !Array.isArray(payload.content)) {
    return { records: [], total: 0 };
  }
  return {
    records: payload.content.map(normalizeRecord),
    total: Number(payload.totalElements) || payload.content.length,
  };
}

export function getAllRecords({ page = 0, size = MAX_PAGE_SIZE } = {}) {
  return apiFetch('/api/records', { params: { page, size } }).then(normalizeList);
}

export function getRecordById(id) {
  return apiFetch(`/api/records/${id}`).then(normalizeRecord);
}

// Build a search request for GET /api/records/search. Only populated
// parameters are sent so the URL stays clean for the required filters.
export function searchRecords({
  q,
  category,
  year,
  status,
  department,
  page = 0,
  size = MAX_PAGE_SIZE,
} = {}) {
  const params = {};
  const query = String(q || '').trim();
  if (query) params.q = query;
  const apiCategory = categoryToApi(category);
  if (apiCategory && apiCategory !== 'All') params.category = apiCategory;
  if (year && year !== 'All Years' && year !== '') params.year = Number(year);
  if (status && status !== 'All') params.status = status;
  if (department && department !== 'All Departments') params.department = department;
  params.page = page;
  params.size = size;
  return apiFetch('/api/records/search', { params }).then(normalizeList);
}

export function getCategories() {
  return apiFetch('/api/categories').then((items) =>
    (Array.isArray(items) ? items : []).map((item) => ({
      code: item?.code,
      name: item?.name,
    }))
  );
}

export function createRecord(payload) {
  return apiFetch('/api/records', { method: 'POST', body: payload }).then(normalizeRecord);
}

export function updateRecord(id, payload) {
  return apiFetch(`/api/records/${id}`, { method: 'PUT', body: payload }).then(normalizeRecord);
}

export function deleteRecord(id) {
  return apiFetch(`/api/records/${id}`, { method: 'DELETE' });
}

File: javafx-admin/src/main/java/com/heirs/desktop/api/RecordsApi.java
package com.heirs.desktop.api;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.core.type.TypeReference;
import com.heirs.desktop.dto.CategoryResponse;
import com.heirs.desktop.dto.HealthResponse;
import com.heirs.desktop.dto.PagedResponse;
import com.heirs.desktop.dto.RecordResponse;

/**
 * Client for HEIRS Record, Category, and Health REST endpoints.
 */
public class RecordsApi {

    private final ApiClient apiClient;

    public RecordsApi() {
        this(new ApiClient());
    }

    public RecordsApi(ApiClient apiClient) {
        this.apiClient = apiClient;
    }

    public boolean checkHealth() {
        try {
            HealthResponse health = apiClient.get("/api/health", HealthResponse.class);
            return health != null && health.isHealthy();
        } catch (Exception e) {
            return false;
        }
    }

    public RecordResponse createRecord(com.heirs.desktop.dto.RecordRequest request) {
        return apiClient.post("/api/records", request, RecordResponse.class);
    }

    public RecordResponse updateRecord(Long id, com.heirs.desktop.dto.RecordRequest request) {
        requireId(id);
        return apiClient.put("/api/records/" + id, request, RecordResponse.class);
    }

    public void deleteRecord(Long id) {
        requireId(id);
        apiClient.delete("/api/records/" + id);
    }

    private static void requireId(Long id) {
        if (id == null || id <= 0) throw new IllegalArgumentException("A positive record ID is required");
    }

    public PagedResponse<RecordResponse> getAllRecords(int page, int size) {
        String path = "/api/records?page=" + page + "&size=" + size;
        return apiClient.get(path, new TypeReference<PagedResponse<RecordResponse>>() {});
    }

    public RecordResponse getRecordById(Long id) {
        if (id == null) {
            throw new IllegalArgumentException("Record id must not be null");
        }
        return apiClient.get("/api/records/" + id, RecordResponse.class);
    }

    public PagedResponse<RecordResponse> searchRecords(String query,
                                                       String category,
                                                       Integer year,
                                                       String status,
                                                       String department,
                                                       int page,
                                                       int size,
                                                       String sortBy,
                                                       String sortDirection) {
        List<String> queryParams = new ArrayList<>();
        if (query != null && !query.isBlank()) {
            queryParams.add("q=" + encode(query.trim()));
        }
        if (category != null && !category.isBlank() && !"All Categories".equalsIgnoreCase(category)) {
            queryParams.add("category=" + encode(category.trim()));
        }
        if (year != null && year > 0) {
            queryParams.add("year=" + year);
        }
        if (status != null && !status.isBlank() && !"All Statuses".equalsIgnoreCase(status)) {
            queryParams.add("status=" + encode(status.trim()));
        }
        if (department != null && !department.isBlank() && !"All Departments".equalsIgnoreCase(department)) {
            queryParams.add("department=" + encode(department.trim()));
        }
        if (sortBy != null && !sortBy.isBlank()) {
            queryParams.add("sortBy=" + encode(sortBy.trim()));
        }
        if (sortDirection != null && !sortDirection.isBlank()) {
            queryParams.add("sortDirection=" + encode(sortDirection.trim()));
        }
        queryParams.add("page=" + page);
        queryParams.add("size=" + size);

        String path = "/api/records/search?" + String.join("&", queryParams);
        return apiClient.get(path, new TypeReference<PagedResponse<RecordResponse>>() {});
    }

    public List<CategoryResponse> getCategories() {
        return apiClient.get("/api/categories", new TypeReference<List<CategoryResponse>>() {});
    }

    private String encode(String val) {
        return URLEncoder.encode(val, StandardCharsets.UTF_8);
    }
}
